gluster/subvol/cmd/gluster-subvol-provisioner/gluster-subvol-provisioner.go

```diff
@@ -0,0 +1,609 @@
+	var supervolNS, supervolPVCName string
+	gidAllocate := true
+	for k, v := range options.Parameters {
+		switch strings.ToLower(k) {
+		case "namespace":
+			supervolNS = v
+		case "pvc":
+			supervolPVCName = v
+		case "gidmin":
+		// Let allocator handle
+		case "gidmax":
+		// Let allocator handle
+		case "gidallocate":
+			b, err := strconv.ParseBool(v)
+			if err != nil {
+				return nil, fmt.Errorf("invalid value %s for parameter %s: %v", v, k, err)
+			}
+			gidAllocate = b
+		}
+	}
+
+	if supervolPVCName == "" {
+		return nil, fmt.Errorf("pvc is a required options for volume plugin %s", provisionerName)
+	}
+
+	// get the PVC that has been configured
+	supervolPVC, err := p.getPVC(supervolNS, supervolPVCName)
+	if err != nil {
+		return nil, fmt.Errorf("could not find pvc %s in namespace %s", supervolPVCName, supervolNS)
+	}
+
+	err = p.validateProvisionRequirements(options.PVC, supervolPVC)
+	if err != nil {
+		return nil, err
+	}
+
+	// based on the PVC we can get the PV
+	supervolPV, err := p.client.CoreV1().PersistentVolumes().Get(supervolPVC.Spec.VolumeName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("could not find PV for PVC %s/%s", supervolNS, supervolPVCName)
+	}
+
+	// we only support Gluster Volumes for now
+	if supervolPV.Spec.Glusterfs == nil {
+		return nil, fmt.Errorf("the supervol PVC %s/%s needs to be of type Glusterfs", supervolNS, supervolPVCName)
+	}
+
+	// mount the supervolPV
+	mountpoint, err := p.mountPV(supervolNS, supervolPV)
+	if err != nil {
+		return nil, fmt.Errorf("failed to mount PV %s: %s", supervolPV, err)
+	}
+
+	/* TODO:
+	 * - set quota for the size
+	 * - verify that the supervol is large enough
+	volSize := options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	volSizeBytes := volSize.Value()
+	volszInt := int(util.RoundUpToGiB(volSizeBytes))
+	*/
+
+	// in case smartcloning fails, the new PVC should be created, but
+	// without the CloneOf annotation. The assumption is that the origin of
+	// the request to clone can fallback to other cloning/copying in case
+	// the CloneRequest annotation was ignored/unknown.
+
+	// full path of the directory for the new PV
+	destDir := p.makeSubvolPath(mountpoint, options.PVC.Namespace, options.PVC.UID)
+
+	// In case an error occurred during cloning, an empty PVC should be
+	// created. This is the same behaviour as provisioners that do not
+	// provide cloning support. The "k8s.io/CloneOf" annotation should only
+	// get set if cloning was successful.
+	sourcePVCRef, err := p.tryClone(options.PVC, mountpoint, destDir)
+	if err != nil || sourcePVCRef == "" {
+		err = os.MkdirAll(destDir, 0775)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to create subdir for new pvc %s: %s", options.PVC.Name, err)
+		}
+		glog.V(1).Infof("successfully created Gluster Subvol %+v with size and volID", destDir)
+	}
+
+	// TODO: need to os.Chgrp() w/ new gid for all files? Should be done by mount process.
```

Comment:
@nixpanic ok, let me ask this question. :)  Whats the GID for Supervol mounts and subvol mounts ? Is it possible to have different permission bits for supervol and subvol ? If both are same,  does the app pod which mounted the subvol has access or permission to access or operate on supervol?

Reply:
The provisioner creates directories on the supervol with GID=0. Once a subvol PVC gets mounted for a pod, the subdirectory on the supervol gets chgrp'd to whatever the volume plugin in kubelet decides. Because the base directory of the supervol is not mounted on the system where the subvol gets mounted, the pod that uses the subvol can not access other subvol PVCs on the same supervol.